File: other/tutorial/visualization3d.cpp
#include"Population.h"//Single header
using namespace pop;//Population namespace


void visu2DSlice(){
    Mat3UI8 img;
    img.load(POP_PROJECT_SOURCE_DIR+std::string("/image/rock3d.pgm"));
    img.display();//use the arrows  to move in z-axis
}

void visu3DCube(){
    Scene3d scene;
    Mat3UI8 img;
    img.load(POP_PROJECT_SOURCE_DIR+std::string("/image/rock3d.pgm"));
    Visualization::cubeExtruded(scene,img);//add the cube surfaces to the scene
    Visualization::lineCube(scene,img);//add the border red lines to the scene to the scene
    scene.display(false);//display the scene
    waitKey();

}
void visu3DCubeExtrudedWithAxis(){
    Mat3UI8 img;
    img.load(POP_PROJECT_SOURCE_DIR+std::string("/image/rock3d.pgm"));
    Scene3d scene;
    Mat3UI8 extruded(img.getDomain());
    int radius=img.getDomain()(0)/2;
    Vec3I32 x1(0,0,0);
    Vec3I32 x2(img.getDomain());
    ForEachDomain3D(x,extruded){
        if((x-x1).norm(2)<radius||(x-x2).norm(2)<radius)
            extruded(x)=0;
        else
            extruded(x)=255;
    }
    Visualization::cubeExtruded(scene,img,extruded);//add the cube surfaces to the scene
    Visualization::lineCube(scene,img);//add the border red lines to the scene to the scene
    Visualization::axis(scene,40);//add axis
    scene.display(false);//display the scene
    waitKey();
}




void visu3DSlice(){
    Mat3UI8 img;
    img.load(POP_PROJECT_SOURCE_DIR+std::string("/image/rock3d.pgm"));
    Scene3d scene;
    Visualization::plane(scene,img,50,2);
    Visualization::plane(scene,img,50,1);
    Visualization::plane(scene,img,200,0);
    Visualization::lineCube(scene,img);
    scene.display(false);
    waitKey();
}


void visu3DMarchingCube(){

    Mat3UI8 img;
    img.load(POP_PROJECT_SOURCE_DIR+std::string("/image/rock3d.pgm"));
    img = img(Vec3I32(0,0,0),Vec3I32(64,64,64));

    //SEGMENTATION OF THE TREE PHASES
    Mat3UI8 imgfilter= Processing::median(img,2);
    Mat3UI8 grain= Processing::threshold(imgfilter,155);
    Mat3UI8 oil = Processing::threshold(imgfilter,70,110);
    oil = Processing::openingRegionGrowing(oil,2);//To remove the interface artefact
    Mat3UI8 air = Processing::threshold(imgfilter,0,40);
    Mat3UI8 seed = Processing::labelMerge(grain,oil);
    seed = Processing::labelMerge(seed,air);
    Mat3UI8 gradient = Processing::gradientMagnitudeDeriche(img,1.5);
    Mat3UI8 water = Processing::watershed(seed,gradient);

    //THE LABEL GRAIN WITH THE GREY-LEVEL 255*0.75
    grain = Processing::labelFromSingleSeed(water,grain);
    grain=Mat3F64(grain)*0.75;

    //THE LABEL OIL WITH THE GREY-LEVEL 255*0.4
    oil = Processing::labelFromSingleSeed(water,oil);
    oil = Mat3F64(oil)*0.4;

    Mat3UI8 grain_oil = grain+oil;
    Scene3d scene;
    Visualization::marchingCube(scene,grain_oil);//add the marching cube of the grain to the scene
    Visualization::lineCube(scene,grain_oil);//add the border red lines to the scene to the scene
    Visualization::axis(scene);
    scene.display(false);//display the scene
    waitKey();
}

int main(){
    visu2DSlice();
    visu3DCube();
    visu3DCubeExtrudedWithAxis();
    visu3DMarchingCube();
    visu3DSlice();
}

File: other/tutorial/visualization3d.py
import sys
import os
PathPop= ""
if os.path.isfile(PathPop+"population.py")==0:
    print "set the variable PathPop to the path where you compile population, for instance D:\Users/vtariel/Desktop/ANV/Population-build/. This folder must contain population.py"
    sys.exit(-1)
sys.path.append(PathPop)
from population import * 
    

def visu2DSlice() :
    img = Mat3UI8(PathPop+"/image/rock3d.pgm")
    img.display()#use the arrows  to move in z-axis

def visu3DCube():
    scene = Scene3d() 
    img = Mat3UI8(PathPop+"/image/rock3d.pgm")
    visu = Visualization()
    visu.cubeExtruded(scene,img)#add the cube surfaces to the scene
    visu.lineCube(scene,img)#add the border red lines to the scene to the scene
    scene.display(0)#display the scene
    waitKey()

def visu3DSlice():
    scene = Scene3d() 
    img = Mat3UI8(PathPop+"/image/rock3d.pgm")
    visu = Visualization()
    visu.plane(scene,img,50,2)
    visu.plane(scene,img,50,1)
    visu.plane(scene,img,200,0)
    visu.lineCube(scene,img)
    scene.display(0)
    waitKey()

def visu3DMarchingCube():
    img = Mat3UI8()
    img.load(PathPop+"/image/rock3d.pgm")
    img = img(Vec3I32(0,0,0),Vec3I32(64,64,64))
    pde = PDE()
    imgfilter= pde.nonLinearAnisotropicDiffusionDericheFast(img)
    app=Application()
    #app.thresholdSelection(imgfilter)
    proc = Processing()
    grain= proc.threshold(imgfilter,155)
    oil = proc.threshold(imgfilter,70,110)
    oil = proc.openingRegionGrowing(oil,2)
    air = proc.threshold(imgfilter,0,40)
    seed = proc.labelMerge(grain,oil)
    seed = proc.labelMerge(seed,air)
    visu =Visualization()
    #visu.labelForeground(seed,imgfilter).display()
    gradient = proc.gradientMagnitudeDeriche(img,1.5)
    water = proc.watershed(seed,gradient)
    grain = proc.labelFromSingleSeed(water,grain)
    grain=grain/2
    oil = proc.labelFromSingleSeed(water,oil)
    oil = oil/4
    scene = Scene3d()
    visu.marchingCube(scene,grain)
    visu.marchingCube(scene,oil)
    visu.lineCube(scene,img)
    scene.display(0)
    waitKey()
    
def visu3DCubeExtrudedWithAxis():
    scene = Scene3d() 
    img = Mat3UI8()
    img.load(PathPop+"/image/rock3d.pgm")
    img = img(Vec3I32(0,0,0),Vec3I32(64,64,64))
    extruded = Mat3UI8(img.getDomain())
    radius=img.sizeI()/2
    center = Vec3I32(img.getDomain()/2)
    it = img.getIteratorEDomain()
    while it.next():
        if (it.x()-center).norm(2)<radius:
            extruded.setValue(it.x(),100)
    visu = Visualization()
    visu.cubeExtruded(scene,img,extruded)#add the cube surfaces to the scene
    visu.lineCube(scene,img)#add the border red lines to the scene to the scene
    scene.display(0)#display the scene
    waitKey()
        
  
visu2DSlice()
visu3DCube()
visu3DSlice()
visu3DMarchingCube()
visu3DCubeExtrudedWithAxis()

